content/blog/2019-10-23-loop-unrolling-with-bril.md

```diff
@@ -0,0 +1,154 @@
+
+#### Static Analysis Evaluation
+The program is located [here](https://github.com/seanlatias/bril/blob/arrays/examples/loop_unroll.py). To execute the pass, run the following command.
+
+``python loop_unroll.py output.json < input.json``
+
+The pass can successfully unroll the loops in our test programs. For the "double loop" test case, the pass correctly identifies the inner loop and unrolls it. Execution time of the original and unrolled versions are as follows:
+
+| Benchmark | Original (s) | Unrolled (s) |
+|:---------:|:------------:|:------------:|
+|Two consecutive loops|3.530|3.621|
+|Nested loop|3.509|3.533|
+|Vector sum|3.599|3.508|
+|Set vector value|3.473|3.671|
+
+There is no significant performance difference between the unrolled version and the original version. One reason might be 
+because the loop tripcounts in the test programs are all very small. As a result, we modify the vector sum program to have 
+large tripcount (2000) and compare the performance again. To enable this we modify our loop unrolling pass to allow unrolling very large loops. In terms of the sizes of the executables, the unrolled version is 12.7x larger than the original version. The original and unrolled program uses 3.693s and 3.755s, repectively. There is still no significant performance benefit. The possible reason is that the control flow of our test examples is too regular and can be perfectly predicted by the branch predictor in modern processors. Since the processor is pipelined, we do not observe significant performance improvement even after we remove the redundant branches after unrolling. For most benchmarks we actually observe slowdowns, which possibly result from instruction cache misses caused by larger executable sizes. 
+
+#### Dynamic Profiling Evaluation
+The program is located [here](https://github.com/seanlatias/bril/blob/arrays/examples/dyn.py). To execute the pass, run the following command.
+
+``python dyn.py < input.json > output.json``
+
+The pass can run successfully in different test cases including consecutive loops and double loops. In this section we mainly focus on the performance perspective. We construct a vector add program with large trip count numbers (2000), which can benefit loop unrolling and out-of-order execution of modern processors. The original and unrolled program takes 9.111s and 8.341s (measured with linux time utility in usr time) for 10000 execution.
```

Comment:
Any theories about why you see performance improvements with this version and not the other one?